File: yt/yt/flow/examples/python/shuffle/event_reducer.py
"""EventReducer: RowFunction that maintains count in external state."""

from yt.yt.flow.library.python.companion.computation import RowFunction


# [BEGIN event_reducer]
class EventReducer(RowFunction):
    """Counts events per key using external state."""

    def on_message(self, message, output, ctx):
        state = ctx.external_state("/shuffle-state", message)
        builder = state.to_builder()
        builder.set("count", (state.get("count") or 0) + 1)
        state.set(builder.finish())


# [END event_reducer]

File: yt/yt/flow/examples/go/shuffle/event_reducer.go
package main

import (
	"context"

	"go.ytsaurus.tech/yt/go/flow"
)

const (
	shuffleStateName = "/shuffle-state"
	countColumn      = "count"
)

// [BEGIN event_reducer]

type shuffleState struct {
	Count int64 `yson:"count"`
}

type eventReducer struct{}

var _ flow.RowFunction = (*eventReducer)(nil)

func (*eventReducer) OnMessage(
	ctx context.Context,
	rt flow.Runtime,
	msg flow.ExtendedMessage,
	out flow.OutputCollector,
) error {
	state, err := flow.OpenExternalState(rt, shuffleStateName, msg)
	if err != nil {
		return err
	}

	var counter shuffleState
	if _, err := state.ConvertTo(&counter); err != nil {
		return err
	}
	counter.Count++
	return state.ConvertFrom(&counter)
}

// [END event_reducer]
